### ISSUE TITLE
Okx fetch_ohlcv candles/history candles automatic switch issue

### ISSUE THREAD

**Issue by Alfavisioner** [ccxt/ccxt#20756]
### Operating System

Windows 10

### CCXT Version

4.2.9

### Description

Okx states that limit is 300 for live candles, and 100 for history candles
And api may return 1440 live candles

If i query live candles started from 1200 candles back i expect to get 300 live candles, not 100 history ones

since = self.exchange.milliseconds() - 1200 * self.exchange.parse_timeframe('5m')
klines = self.exchange.fetch_ohlcv(symbol='BTCUSDT', timeframe='5m', since=since, limit=300, params=params)

Code returns only 100 results. But has to return 300.

With this code i can get 300 candles only for last 1140 candles (This is 1440 - 300), not for 1440

Why?

Because in okx fetch_ohlcv limit substracted from available 1440 candles. 
It has not to be substracted

### Code
okx:2080-2091
```
           if since is not None:
            now = self.milliseconds()
            difference = now - since
            durationInMilliseconds = duration * 1000
            # if the since timestamp is more than limit candles back in the past
            # additional one bar for max offset to round the current day to UTC
            calc = (1440 - limit - 1) * durationInMilliseconds
            if difference > calc:
                defaultType = 'HistoryCandles'
            startTime = max(since - 1, 0)
            request['before'] = startTime
            request['after'] = self.sum(startTime, durationInMilliseconds * limit)`

```

calc = (1440 - limit - 1) * durationInMilliseconds
has to be replaced to 
calc = 1439 * durationInMilliseconds

**Comment by Dan-krm** [comment]
Hey @Alfavisioner I think your question might be answered by the comment in the code snippet you posted:
`# if the since timestamp is more than limit candles back in the past` then the history candles with a limit of 100 are used, based on that I don't think it makes sense to remove limit from calc because then I'm not sure if the whole `if difference > calc` segment would work properly.

Here is my response from your example fetch_ohlcv call:
```
okx.fetchOHLCV (BTC/USDT, 5m, 1704955844524, 300)
2024-01-11T06:56:44.919Z iteration 0 passed in 396 ms

1704956100000 | 45767.6 | 45900 | 45765.8 | 45765.8 | 2.97980933
1 objects
```
And your proposed changes return the same result for me so I'm not sure what I'm missing.

If I'm misunderstanding your issue please elaborate on your **Description** section in a clear way, I found it a bit hard to follow what you were trying to say

You can also review this PR where the change that you are debating was introduced and see if the logic stands https://github.com/ccxt/ccxt/pull/18264

Thanks, Dan

**Comment by Alfavisioner** [comment]
Okx exchange use 2 modes to return candles
1. Live mode for last 1440 candles and with limit 300 and maximim 40 requests per 2 seconds
2. History mode with limit 100 and maximim 20 requests per 2 seconds

Look at cases:
1.
```
from ccxt import okx
ex = okx()

hourInMs = 3600 * 1000
since = ex.milliseconds() - 1439 * hourInMs
klines = ex.fetch_ohlcv('BTC/USDT', '1h', since, limit= 300)
print(len(klines))
```

### Output
100
Has to be 300

Case 1 queries klines 1440 periods back. We deduct 1 to take into account current unclosed candle.
According to okex documentation it has to return 300 candles.
It returns only 100 instead of 300

2.
```
from ccxt import okx
ex = okx()

hourInMs = 3600 * 1000
since = ex.milliseconds() - 1200 * hourInMs
klines = ex.fetch_ohlcv('BTC/USDT', '1h', since, limit= 300)
print(len(klines))
```

### Output
100
Has to be 300

Case 2 query 1200 klines back
Returns only 100 instead of 300

3.
```
from ccxt import okx
ex = okx()

hourInMs = 3600 * 1000
since = ex.milliseconds() - 1000 * hourInMs
klines = ex.fetch_ohlcv('BTC/USDT', '1h', since, limit= 300)
print(len(klines))
```

### Output
300
All is ok

Case 3 query 1000 klines back
Returns 300 as expected

Why:
Because of deducted limit in calc variable. Code just switched from Candles to HisitoryCandles earlier than necessary:
calc = (1440 - limit - 1) * durationInMilliseconds
If you deduct limit variable from 1440, you just move minimum possible Candles timestamp to 1140 in out case  

First of all i investigated issue: https://github.com/ccxt/ccxt/pull/18264

Found that it is posible 3 cases:
a) _before_ and _after_ is left on timeline camparing to 1440 border. HistoryCandles mode has to be used:
b) _before_ and _after_ is right on timeline camparing to 1440 border. Candles mode has to be used:
c) _before_ is left, _after_ is right   (case https://github.com/ccxt/ccxt/pull/18264). HistoryCandles mode has to be used:


My proposed solution:

To be frankly, calc variable is non descriptive and difference variable confusing too:
Propose to use history_border variable instead:
historyBorder = now - ((1440 - 1) * durationInMilliseconds)

Propose to change code from:
```
      now = self.milliseconds()
      difference = now - since
      durationInMilliseconds = duration * 1000
      # if the since timestamp is more than limit candles back in the past
      # additional one bar for max offset to round the current day to UTC
      calc = (1440 - limit - 1) * durationInMilliseconds
      if difference > calc:
          defaultType = 'HistoryCandles'
      startTime = max(since - 1, 0)
      request['before'] = startTime
      request['after'] = self.sum(startTime, durationInMilliseconds * limit)
```

to:

```
    now = self.milliseconds()
    durationInMilliseconds = duration * 1000
    # if the since timestamp is more than limit candles back in the past
    # additional one bar for max offset to round the current day to UTC
    historyBorder = now - ((1440 - 1) * durationInMilliseconds)
    startTime = max(since - 1, 0)
    if startTime >= historyBorder:
        defaultType = 'Candles'
    else:
        defaultType = 'HistoryCandles'
    request['before'] = startTime
    request['after'] = self.sum(startTime, durationInMilliseconds * limit)
```

**Comment by Dan-krm** [comment]
@alfavisioner thanks for the thorough explanation, your first example test was failing for me unless I set:
`since = this.milliseconds() - 1438 * hourInMs;` it passes if I removed the `-1` from either `historyBorder` or `startTime`, in the pull request I opened, I removed the `-1`  from both `historyBorder` and `since` and simplified the code a bit so now we can run:
```
from ccxt import okx
ex = okx()

hourInMs = 3600 * 1000
since = ex.milliseconds() - 1439 * hourInMs
klines = ex.fetch_ohlcv('BTC/USDT', '1h', since, limit= 300)
print(len(klines))

### Output
300 candles
```
and
```
from ccxt import okx
ex = okx()

hourInMs = 3600 * 1000
since = ex.milliseconds() - 1440 * hourInMs
klines = ex.fetch_ohlcv('BTC/USDT', '1h', since, limit= 300)
print(len(klines))

### Output
100 candles
```
Let me know if you notice any flaws in this PR: #20802
